File: cpp/src/yaml.cpp
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause
//
// YAML check loader implementation.
//
#include <aletheia/yaml.hpp>

#include "detail/loader_utils.hpp"

#include <yaml-cpp/yaml.h>

#include <chrono>
#include <cstdint>
#include <expected>
#include <filesystem>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

namespace aletheia {

// ---------------------------------------------------------------------------
// YAML field extractors with error context
// ---------------------------------------------------------------------------

// The refusal every field reader below throws, for a key that is absent or
// that names a node of another kind; the wording is shared with the Python
// loader, so it has one owner here.
static auto missing_or_invalid(const std::string& ctx, const std::string& key,
                               std::string_view expected) -> std::runtime_error {
    return std::runtime_error(ctx + ": missing or invalid '" + key + "' (expected " +
                              std::string{expected} + ")");
}

// The child a key names, refused in the loader's words when absent. Every
// reader takes its child through here, so an undefined node never reaches a
// kind test, whose refusal would be yaml-cpp's rather than the loader's.
static auto require_child(const YAML::Node& node, const std::string& key, const std::string& ctx,
                          std::string_view expected) -> YAML::Node {
    auto const child = node[key];
    if (!child)
        throw missing_or_invalid(ctx, key, expected);
    return child;
}

static auto get_str(const YAML::Node& node, const std::string& key, const std::string& ctx)
    -> std::string {
    auto const child = require_child(node, key, ctx, "string");
    if (!child.IsScalar())
        throw missing_or_invalid(ctx, key, "string");
    return child.as<std::string>();
}

// Read a numeric scalar as an EXACT Rational via the kernel decimal SSOT
// (`Rational::from_decimal`) — the float principle: no float ever materialises.
// YAML preserves the original scalar text, so the literal "11.5" is handed to
// the kernel verbatim (→ 23/2) instead of round-tripping through a double.
// RTS-gated: an FfiBackend must be live first. A kernel refusal of the literal
// is the document's own defect, re-thrown with the check's context prefixed
// because the kernel knows the literal and not the position; a refusal that is
// not about the literal keeps its own kind all the way out of the loader.
static auto get_decimal(const YAML::Node& node, const std::string& key, const std::string& ctx)
    -> Rational {
    auto const child = require_child(node, key, ctx, "number");
    if (!child.IsScalar())
        throw missing_or_invalid(ctx, key, "number");
    // Reject booleans: yaml-cpp parses "true"/"false" as scalars too, and the
    // kernel grammar would otherwise reject them with a less specific message.
    auto const raw = child.as<std::string>();
    if (raw == "true" || raw == "false" || raw == "TRUE" || raw == "FALSE" || raw == "True" ||
        raw == "False")
        throw missing_or_invalid(ctx, key, "number");
    try {
        return Rational::from_decimal(raw);
    } catch (const AletheiaException& ex) {
        if (ex.kind() != ErrorKind::Validation)
            throw; // runtime-down / ABI faults are not properties of the document
        throw std::runtime_error(ctx + ": invalid '" + key + "': " + ex.what());
    }
}

static auto get_int(const YAML::Node& node, const std::string& key, const std::string& ctx)
    -> std::int64_t {
    auto const child = require_child(node, key, ctx, "integer");
    // A non-scalar fails the conversion below with the same refusal.
    try {
        return child.as<std::int64_t>();
    } catch (const YAML::BadConversion&) {
        throw missing_or_invalid(ctx, key, "integer");
    }
}

static auto get_map(const YAML::Node& node, const std::string& key, const std::string& ctx)
    -> YAML::Node {
    auto const child = require_child(node, key, ctx, "mapping");
    if (!child.IsMap())
        throw missing_or_invalid(ctx, key, "mapping");
    return child;
}

// ---------------------------------------------------------------------------
// Check name extraction
// ---------------------------------------------------------------------------

static auto check_name(const YAML::Node& entry) -> std::string {
    auto const name_node = entry["name"];
    if (name_node && name_node.IsScalar())
        return name_node.as<std::string>();
    return "<unnamed>";
}

static auto ctx(const std::string& name) -> std::string {
    return "Check '" + name + "'";
}

// ---------------------------------------------------------------------------
// Simple check parser
// ---------------------------------------------------------------------------

static auto parse_simple_check(const YAML::Node& entry, const std::string& name) -> CheckResult {
    auto const condition = get_str(entry, "condition", ctx(name));
    auto const signal = get_str(entry, "signal", ctx(name));

    if (!detail::is_simple_condition(condition))
        throw std::runtime_error(ctx(name) + ": unknown condition '" + condition + "'");

    if (detail::is_simple_value_condition(condition)) {
        if (!entry["value"])
            throw std::runtime_error(ctx(name) + ": condition '" + condition +
                                     "' requires 'value'");
        auto const value = PhysicalValue{get_decimal(entry, "value", ctx(name))};
        return detail::dispatch_simple(signal, condition, value);
    }

    if (detail::is_simple_range_condition(condition)) {
        if (!entry["min"] || !entry["max"])
            throw std::runtime_error(ctx(name) + ": condition '" + condition +
                                     "' requires 'min' and 'max'");
        auto const lo = PhysicalValue{get_decimal(entry, "min", ctx(name))};
        auto const hi = PhysicalValue{get_decimal(entry, "max", ctx(name))};
        return check::signal(signal).stays_between(lo, hi);
    }

    if (detail::is_simple_settles_condition(condition)) {
        if (!entry["min"] || !entry["max"])
            throw std::runtime_error(ctx(name) +
                                     ": condition 'settles_between' requires 'min' and 'max'");
        if (!entry["within_ms"])
            throw std::runtime_error(ctx(name) +
                                     ": condition 'settles_between' requires 'within_ms'");
        auto const lo = PhysicalValue{get_decimal(entry, "min", ctx(name))};
        auto const hi = PhysicalValue{get_decimal(entry, "max", ctx(name))};
        auto const ms = std::chrono::milliseconds{get_int(entry, "within_ms", ctx(name))};
        return check::signal(signal).settles_between(lo, hi).within(ms);
    }

    // equals
    if (!entry["value"])
        throw std::runtime_error(ctx(name) + ": condition 'equals' requires 'value'");
    auto const value = PhysicalValue{get_decimal(entry, "value", ctx(name))};
    return check::signal(signal).equals(value).always();
}

// ---------------------------------------------------------------------------
// When/Then check parser
// ---------------------------------------------------------------------------

static auto parse_when_then_check(const YAML::Node& entry, const std::string& name) -> CheckResult {
    if (!entry["then"])
        throw std::runtime_error(ctx(name) + ": must have 'signal' or 'when'/'then'");
    if (!entry["within_ms"])
        throw std::runtime_error(ctx(name) + ": when/then checks require 'within_ms'");

    auto const when = get_map(entry, "when", ctx(name));
    auto then = get_map(entry, "then", ctx(name));
    auto const within_ms = std::chrono::milliseconds{get_int(entry, "within_ms", ctx(name))};

    // When clause
    auto const when_cond = get_str(when, "condition", ctx(name));
    if (!detail::is_when_condition(when_cond))
        throw std::runtime_error(ctx(name) + ": unknown when condition '" + when_cond + "'");

    auto const when_signal = get_str(when, "signal", ctx(name));
    auto const when_value = PhysicalValue{get_decimal(when, "value", ctx(name))};
    auto const when_builder = check::when(when_signal);
    auto const when_result = detail::dispatch_when(when_builder, when_cond, when_value);

    // Then clause
    auto const then_cond = get_str(then, "condition", ctx(name));
    // The word is held to the vocabulary by taking its slots: one lookup
    // answers both whether the obligation is known and what it reads.
    auto const slots = detail::then_slots(then_cond);
    if (!slots)
        throw std::runtime_error(ctx(name) + ": unknown then condition '" + then_cond + "'");

    auto const then_signal = get_str(then, "signal", ctx(name));
    auto const then_builder = when_result.then(then_signal);

    // Which keys the obligation reads is the vocabulary's business, not this
    // loader's; which keys they are, and what to say when one is missing, is
    // this loader's. Only the slots the obligation reads are handed over.
    detail::ThenSlotValues read;
    switch (*slots) {
    case detail::ThenSlots::Value:
        read.emplace("value", PhysicalValue{get_decimal(then, "value", ctx(name))});
        break;
    case detail::ThenSlots::Range:
        if (!then["min"] || !then["max"])
            throw std::runtime_error(ctx(name) + ": then condition '" + then_cond +
                                     "' requires 'min' and 'max'");
        read.emplace("lo", PhysicalValue{get_decimal(then, "min", ctx(name))});
        read.emplace("hi", PhysicalValue{get_decimal(then, "max", ctx(name))});
        break;
    }
    return detail::dispatch_then(then_builder, then_cond, read, within_ms);
}

// ---------------------------------------------------------------------------
// Single check entry parser
// ---------------------------------------------------------------------------

static auto parse_check(const YAML::Node& entry) -> CheckResult {
    auto name = check_name(entry);

    auto result = [&] {
        if (entry["when"])
            return parse_when_then_check(entry, name);
        if (entry["signal"])
            return parse_simple_check(entry, name);
        throw std::runtime_error(ctx(name) + ": must have 'signal' or 'when'/'then'");
    }();

    // Apply metadata
    auto const name_node = entry["name"];
    if (name_node && name_node.IsScalar())
        result.named(name_node.as<std::string>());
    auto const sev_node = entry["severity"];
    if (sev_node && sev_node.IsScalar())
        result.severity(sev_node.as<std::string>());

    return result;
}

// ---------------------------------------------------------------------------
// Top-level YAML parser
// ---------------------------------------------------------------------------

static auto parse_yaml_checks(const YAML::Node& root) -> Result<std::vector<CheckResult>> {
    if (!root.IsMap() || !root["checks"])
        return std::unexpected(
            AletheiaError{ErrorKind::Validation, "YAML must contain a 'checks' list"});

    auto const checks_node = root["checks"];
    if (!checks_node.IsSequence())
        return std::unexpected(
            AletheiaError{ErrorKind::Validation, "YAML must contain a 'checks' list"});

    std::vector<CheckResult> results;
    for (auto const& entry : checks_node) {
        if (!entry.IsMap())
            return std::unexpected(
                AletheiaError{ErrorKind::Validation, "Each check must be a YAML mapping"});
        try {
            results.push_back(parse_check(entry));
        } catch (const AletheiaException& ex) {
            // A kernel or runtime failure keeps its kind; only the document's
            // own defect is a Validation error.
            return std::unexpected(ex.error());
        } catch (const std::runtime_error& ex) {
            return std::unexpected(AletheiaError{ErrorKind::Validation, ex.what()});
        }
    }
    return results;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

auto load_checks_from_yaml(const std::filesystem::path& path) -> Result<std::vector<CheckResult>> {
    // Reject symlinks and cap the raw size before handing the path to yaml-cpp.
    // YAML has no compressed container, so the uncompressed-size walk the .xlsx
    // loader does over its ZIP entries has no counterpart here.
    if (auto v = detail::validate_loader_path(path, "YAML"); !v)
        return std::unexpected(v.error());
    if (auto v = detail::check_file_size_bound(path); !v)
        return std::unexpected(v.error());

    try {
        auto const root = YAML::LoadFile(path.string());
        return parse_yaml_checks(root);
    } catch (const YAML::Exception& ex) {
        return std::unexpected(AletheiaError{ErrorKind::Validation, std::string(ex.what())});
    }
}

auto load_checks_from_yaml_string(std::string_view yaml) -> Result<std::vector<CheckResult>> {
    // Bound the in-memory input before the YAML::Load copy/parse, mirroring the
    // file loader's check_file_size_bound (and Go/Rust, which bound their inline
    // YAML loaders).  AGENTS.md trust boundary: every check loader, file or
    // inline, caps input at max_dbc_text_bytes.
    if (auto v = detail::check_input_size_bound(yaml.size()); !v)
        return std::unexpected(v.error());
    try {
        auto const root = YAML::Load(std::string(yaml));
        return parse_yaml_checks(root);
    } catch (const YAML::Exception& ex) {
        return std::unexpected(AletheiaError{ErrorKind::Validation, std::string(ex.what())});
    }
}

} // namespace aletheia

File: rust/src/yaml.rs
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause

//! YAML check loader (feature `yaml`, on by default).
//!
//! Loads a set of named [`Check`]s from a YAML document whose schema is shared
//! with the Python (`load_checks`) and Go (`LoadChecksFromYAML`) loaders, so a
//! check file authored for one binding loads identically here. The document is a
//! mapping with a single `checks:` list; each entry is either a single-signal
//! check (`signal` + `condition` + operands) or a causal `when`/`then` check
//! (with a top-level `within_ms`). Decimal values are parsed EXACTLY from their
//! original text via the kernel [`Rational::from_decimal`] (the cross-binding
//! single source of truth), so an out-of-grammar or overflowing value fails
//! rather than silently clamping — and loading a check file with decimal operands
//! requires a live backend (decimal parsing is RTS-gated). Unknown keys are
//! ignored (matching the peers).
//!
//! Note: a decimal operand renders in a check's *description* as the exact reduced
//! fraction (`11.5` → `"11.5"`'s value shows as `23/2`), matching the Rust check
//! DSL; the peers' loaders print the original decimal. This is display-only — the
//! compiled formula and wire value are identical across bindings.

use std::path::Path;

use yaml_rust2::{Yaml, YamlLoader};

use crate::check::{self, Check};
use crate::error::Error;
use crate::types::Rational;

/// The shared 64 MiB input-length cap (Python `MAX_DBC_TEXT_BYTES`, Go
/// `MaxDBCTextBytes`). Applied at this trust boundary so an adversarial check
/// file cannot force unbounded allocation / parse time before it is rejected.
const MAX_INPUT_BYTES: usize = 64 * 1024 * 1024;

fn invalid(msg: impl Into<String>) -> Error {
    Error::Validation(msg.into())
}

/// Whether a node is present (a missing mapping key yields [`Yaml::BadValue`]).
fn present(y: &Yaml) -> bool {
    !matches!(y, Yaml::BadValue)
}

/// Load checks from inline YAML content.
///
/// Rejects input larger than the shared 64 MiB bound (`MAX_INPUT_BYTES`).
///
/// # Errors
/// [`Error::Validation`] if the input exceeds the size bound, the document is
/// malformed, a check is missing a required field, a condition keyword is
/// unknown, or a numeric value is out of the decimal grammar or overflows `i64`
/// (see [`Rational::from_decimal`]); [`Error::RtsNotInitialized`] if no backend
/// is up (decimal parsing is RTS-gated).
pub fn load_checks_from_yaml(content: &str) -> Result<Vec<Check>, Error> {
    load_checks_within(content, MAX_INPUT_BYTES)
}

/// Inline loader with an injectable size bound (the public entry point fixes it
/// to [`MAX_INPUT_BYTES`]; tests use a small bound to exercise the boundary
/// without allocating 64 MiB).
fn load_checks_within(content: &str, limit: usize) -> Result<Vec<Check>, Error> {
    if content.len() > limit {
        return Err(invalid(format!(
            "YAML check input exceeds the {limit}-byte input bound ({} bytes)",
            content.len()
        )));
    }
    let docs =
        YamlLoader::load_from_str(content).map_err(|e| invalid(format!("invalid YAML: {e}")))?;
    let doc = docs
        .first()
        .ok_or_else(|| invalid("YAML must contain a 'checks' list"))?;
    let entries = doc["checks"]
        .as_vec()
        .ok_or_else(|| invalid("YAML must contain a 'checks' list"))?;
    entries.iter().map(parse_check).collect()
}

/// Load checks from a YAML file.
///
/// Mirrors the Go and Python loaders' trust-boundary hardening: the path is
/// rejected if it is a symbolic link or a non-regular file, and the file's size
/// is checked against the shared 64 MiB bound (`MAX_INPUT_BYTES`) **before** it
/// is read, so an adversarial path cannot trigger a huge read.
///
/// # Errors
/// [`Error::Validation`] if the path is a symlink / non-regular file, the file
/// exceeds the size bound, an I/O error occurs, or the document is malformed (see
/// [`load_checks_from_yaml`]).
pub fn load_checks_from_yaml_file(path: impl AsRef<Path>) -> Result<Vec<Check>, Error> {
    read_checks_file(path.as_ref(), MAX_INPUT_BYTES)
}

/// File loader with an injectable size bound (see [`load_checks_within`]).
fn read_checks_file(path: &Path, limit: usize) -> Result<Vec<Check>, Error> {
    // symlink_metadata is an lstat — it does NOT follow a symlink, so a symlinked
    // path is detected here rather than silently dereferenced.
    let meta = std::fs::symlink_metadata(path)
        .map_err(|e| invalid(format!("cannot stat {}: {e}", path.display())))?;
    let file_type = meta.file_type();
    if file_type.is_symlink() {
        return Err(invalid(format!(
            "refusing to load {}: symbolic link (resolve the link and pass the real path)",
            path.display()
        )));
    }
    if !file_type.is_file() {
        return Err(invalid(format!(
            "refusing to load {}: not a regular file",
            path.display()
        )));
    }
    if meta.len() > limit as u64 {
        return Err(invalid(format!(
            "YAML check file {} exceeds the {limit}-byte input bound ({} bytes)",
            path.display(),
            meta.len()
        )));
    }
    let content = std::fs::read_to_string(path)
        .map_err(|e| invalid(format!("cannot read {}: {e}", path.display())))?;
    load_checks_within(&content, limit)
}

fn parse_check(entry: &Yaml) -> Result<Check, Error> {
    let name = entry["name"].as_str().unwrap_or("");
    let ctx = if name.is_empty() { "<unnamed>" } else { name };

    let check = if present(&entry["when"]) {
        parse_when_then(entry, ctx)?
    } else if present(&entry["signal"]) {
        parse_simple(entry, ctx)?
    } else {
        return Err(invalid(format!(
            "check '{ctx}': must have 'signal' or 'when'/'then'"
        )));
    };

    // Metadata is applied only when non-empty (matching the peers).
    let check = if name.is_empty() {
        check
    } else {
        check.named(name)
    };
    let severity = entry["severity"].as_str().unwrap_or("");
    let check = if severity.is_empty() {
        check
    } else {
        check.with_severity(severity)
    };
    Ok(check)
}

fn parse_simple(entry: &Yaml, ctx: &str) -> Result<Check, Error> {
    let signal = entry["signal"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': 'signal' must be a string")))?;
    let condition = entry["condition"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': missing 'condition'")))?;
    let sig = check::signal(signal);
    match condition {
        "never_exceeds" => Ok(sig.never_exceeds(number(entry, "value", ctx, condition)?)),
        "never_below" => Ok(sig.never_below(number(entry, "value", ctx, condition)?)),
        "never_equals" => Ok(sig.never_equals(number(entry, "value", ctx, condition)?)),
        "equals" => Ok(sig.equals(number(entry, "value", ctx, condition)?).always()),
        "stays_between" => {
            let (lo, hi) = range(entry, ctx, condition)?;
            sig.stays_between(lo, hi)
        }
        "settles_between" => {
            let (lo, hi) = range(entry, ctx, condition)?;
            let ms = within_ms(entry, ctx, condition)?;
            sig.settles_between(lo, hi).within(ms)
        }
        other => Err(invalid(format!(
            "check '{ctx}': unknown condition '{other}'"
        ))),
    }
}

fn parse_when_then(entry: &Yaml, ctx: &str) -> Result<Check, Error> {
    let when = &entry["when"];
    let then = &entry["then"];
    if !present(then) {
        return Err(invalid(format!(
            "check '{ctx}': must have 'signal' or 'when'/'then'"
        )));
    }
    // within_ms lives at the entry's top level and is mandatory for when/then.
    let ms = within_ms(entry, ctx, "when/then")?;

    let when_signal = when["signal"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': when clause requires 'signal'")))?;
    let when_cond = when["condition"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': when clause requires 'condition'")))?;
    let trigger = check::when(when_signal);
    let when_condition = match when_cond {
        "exceeds" => trigger.exceeds(number(when, "value", ctx, when_cond)?),
        "equals" => trigger.equals(number(when, "value", ctx, when_cond)?),
        "drops_below" => trigger.drops_below(number(when, "value", ctx, when_cond)?),
        other => {
            return Err(invalid(format!(
                "check '{ctx}': unknown when condition '{other}'"
            )))
        }
    };

    let then_signal = then["signal"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': then clause requires 'signal'")))?;
    let then_cond = then["condition"]
        .as_str()
        .ok_or_else(|| invalid(format!("check '{ctx}': then clause requires 'condition'")))?;
    let response = when_condition.then(then_signal);
    let then_condition = match then_cond {
        "equals" => response.equals(number(then, "value", ctx, then_cond)?),
        "exceeds" => response.exceeds(number(then, "value", ctx, then_cond)?),
        "stays_between" => {
            let (lo, hi) = range(then, ctx, then_cond)?;
            response.stays_between(lo, hi)
        }
        other => {
            return Err(invalid(format!(
                "check '{ctx}': unknown then condition '{other}'"
            )))
        }
    };
    then_condition.within(ms)
}

/// Extract a numeric operand as a [`Rational`] — an integer scalar becomes `n/1`,
/// a decimal scalar is parsed EXACTLY from its original text via the kernel
/// [`Rational::from_decimal`] (no float ever materialises; `yaml_rust2` keeps the
/// literal as [`Yaml::Real`]). A boolean or non-number is rejected, as is a
/// missing key. Because decimal parsing is RTS-gated, loading a check file with
/// decimal operands requires a live backend.
fn number(node: &Yaml, key: &str, ctx: &str, cond: &str) -> Result<Rational, Error> {
    match &node[key] {
        Yaml::Integer(i) => Ok(Rational::integer(*i)),
        Yaml::Real(s) => Rational::from_decimal(s),
        _ => Err(invalid(format!(
            "check '{ctx}': condition '{cond}' requires '{key}'"
        ))),
    }
}

/// Extract the `min` and `max` operands of a range condition (both required).
fn range(node: &Yaml, ctx: &str, cond: &str) -> Result<(Rational, Rational), Error> {
    if !present(&node["min"]) || !present(&node["max"]) {
        return Err(invalid(format!(
            "check '{ctx}': condition '{cond}' requires 'min' and 'max'"
        )));
    }
    Ok((
        number(node, "min", ctx, cond)?,
        number(node, "max", ctx, cond)?,
    ))
}

/// Extract the required integer-millisecond `within_ms` operand (non-negative).
fn within_ms(node: &Yaml, ctx: &str, cond: &str) -> Result<u64, Error> {
    let v = &node["within_ms"];
    if !present(v) {
        return Err(invalid(format!(
            "check '{ctx}': condition '{cond}' requires 'within_ms'"
        )));
    }
    let ms = v
        .as_i64()
        .ok_or_else(|| invalid(format!("check '{ctx}': 'within_ms' must be an integer")))?;
    u64::try_from(ms)
        .map_err(|_| invalid(format!("check '{ctx}': 'within_ms' must be non-negative")))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Rational;

    fn one(yaml: &str) -> Check {
        let mut checks = load_checks_from_yaml(yaml).expect("load");
        assert_eq!(checks.len(), 1, "expected exactly one check");
        checks.pop().unwrap()
    }

    /// Bring the GHC RTS up: decimal operands are parsed via the kernel decimal
    /// SSOT (`Rational::from_decimal`), which is RTS-gated, so any test that loads
    /// a check with a decimal value needs a live backend first. `Client::new()` is
    /// the sole RTS initialiser; the process-global one-shot latch keeps it up.
    fn test_backend() -> crate::Client {
        crate::Client::new().expect(
            "init GHC RTS for decimal-loader test (is ALETHEIA_LIB set to a built libaletheia-ffi.so?)",
        )
    }

    #[test]
    fn simple_conditions_match_the_builders() {
        // Each loaded check must compile to the same formula as the DSL builder.
        assert_eq!(
            one("checks:\n  - signal: Speed\n    condition: never_exceeds\n    value: 220\n")
                .formula(),
            check::signal("Speed").never_exceeds(220).formula()
        );
        assert_eq!(
            one("checks:\n  - signal: V\n    condition: never_below\n    value: 5\n").formula(),
            check::signal("V").never_below(5).formula()
        );
        assert_eq!(
            one("checks:\n  - signal: E\n    condition: never_equals\n    value: 255\n").formula(),
            check::signal("E").never_equals(255).formula()
        );
        assert_eq!(
            one("checks:\n  - signal: G\n    condition: equals\n    value: 0\n").formula(),
            check::signal("G").equals(0).always().formula()
        );
        assert_eq!(
            one("checks:\n  - signal: T\n    condition: stays_between\n    min: 60\n    max: 80\n")
                .formula(),
            check::signal("T").stays_between(60, 80).unwrap().formula()
        );
        assert_eq!(
            one("checks:\n  - signal: T\n    condition: settles_between\n    min: 60\n    max: 80\n    within_ms: 500\n")
                .formula(),
            check::signal("T")
                .settles_between(60, 80)
                .within(500)
                .unwrap()
                .formula()
        );
    }

    #[test]
    fn when_then_matches_the_builder() {
        let c = one(concat!(
            "checks:\n",
            "  - when:\n      signal: Brake\n      condition: exceeds\n      value: 50\n",
            "    then:\n      signal: Light\n      condition: equals\n      value: 1\n",
            "    within_ms: 100\n",
        ));
        assert_eq!(
            c.formula(),
            check::when("Brake")
                .exceeds(50)
                .then("Light")
                .equals(1)
                .within(100)
                .unwrap()
                .formula()
        );
        assert_eq!(c.signal_name(), "Light");
    }

    #[test]
    fn decimal_value_uses_the_shared_conversion() {
        // "100.5" parses EXACTLY to 201/2 via the kernel decimal SSOT — the same
        // Rational every binding produces. Decimal parsing is RTS-gated.
        let _rts = test_backend();
        let c = one("checks:\n  - signal: S\n    condition: never_exceeds\n    value: 100.5\n");
        let expected = check::signal("S").never_exceeds(Rational::new(201, 2).unwrap());
        assert_eq!(c.formula(), expected.formula());
    }

    #[test]
    fn metadata_name_and_severity_are_applied() {
        let c = one(concat!(
            "checks:\n  - name: speed-limit\n    severity: critical\n",
            "    signal: Speed\n    condition: never_exceeds\n    value: 220\n",
        ));
        assert_eq!(c.name(), "speed-limit");
        assert_eq!(c.severity(), "critical");
    }

    #[test]
    fn multiple_checks_and_unknown_keys_ignored() {
        let checks = load_checks_from_yaml(concat!(
            "checks:\n",
            "  - signal: A\n    condition: never_exceeds\n    value: 1\n    bogus_key: 7\n",
            "  - signal: B\n    condition: never_below\n    value: 2\n",
        ))
        .expect("load");
        assert_eq!(checks.len(), 2);
    }

    #[test]
    fn errors_on_malformed_documents() {
        // No checks list.
        assert!(load_checks_from_yaml("not_checks: []\n").is_err());
        // Unknown condition.
        assert!(load_checks_from_yaml(
            "checks:\n  - signal: S\n    condition: bogus\n    value: 1\n"
        )
        .is_err());
        // Missing value.
        assert!(
            load_checks_from_yaml("checks:\n  - signal: S\n    condition: never_exceeds\n")
                .is_err()
        );
        // when/then without within_ms.
        assert!(load_checks_from_yaml(concat!(
            "checks:\n  - when:\n      signal: A\n      condition: exceeds\n      value: 1\n",
            "    then:\n      signal: B\n      condition: equals\n      value: 1\n",
        ))
        .is_err());
        // Inverted range.
        assert!(load_checks_from_yaml(
            "checks:\n  - signal: S\n    condition: stays_between\n    min: 10\n    max: 0\n"
        )
        .is_err());
        // Neither signal nor when.
        assert!(load_checks_from_yaml("checks:\n  - severity: warning\n").is_err());
    }

    #[test]
    fn rejects_non_finite_and_overflowing_values() {
        // `.nan` / `.inf` are out of the kernel decimal grammar; the 20-digit
        // integer part overflows the i64 wire range. The loader rejects all three
        // (never clamps) — matching every binding's decimal SSOT. RTS-gated.
        let _rts = test_backend();
        assert!(load_checks_from_yaml(
            "checks:\n  - signal: S\n    condition: never_exceeds\n    value: .nan\n"
        )
        .is_err());
        assert!(load_checks_from_yaml(
            "checks:\n  - signal: S\n    condition: never_exceeds\n    value: .inf\n"
        )
        .is_err());
        assert!(load_checks_from_yaml(
            "checks:\n  - signal: S\n    condition: never_exceeds\n    value: 99999999999999999999.5\n"
        )
        .is_err());
    }

    // --- Trust-boundary hardening (parity with the Go / Python loaders) -------

    use std::path::PathBuf;

    const SAMPLE: &str = "checks:\n  - signal: S\n    condition: never_exceeds\n    value: 1\n";

    /// A temp path that removes its file on drop.
    struct TempFile(PathBuf);
    impl Drop for TempFile {
        fn drop(&mut self) {
            let _ = std::fs::remove_file(&self.0);
        }
    }
    fn temp_path(tag: &str) -> PathBuf {
        let mut p = std::env::temp_dir();
        p.push(format!("aletheia_yaml_{}_{tag}", std::process::id()));
        p
    }

    #[test]
    fn inline_input_bound_is_enforced() {
        // A small injectable bound exercises the boundary without a 64 MiB alloc;
        // the public entry point fixes the real bound at MAX_INPUT_BYTES.
        assert!(load_checks_within(SAMPLE, 8).is_err());
        assert!(load_checks_within(SAMPLE, SAMPLE.len()).is_ok());
    }

    #[test]
    fn file_loads_and_size_bound_is_enforced() {
        let path = temp_path("ok");
        let _guard = TempFile(path.clone());
        std::fs::write(&path, SAMPLE).expect("write temp");
        assert_eq!(read_checks_file(&path, MAX_INPUT_BYTES).unwrap().len(), 1);
        // Rejected (before reading) when the file exceeds a small bound.
        assert!(read_checks_file(&path, 8).is_err());
    }

    #[test]
    fn rejects_non_regular_file() {
        let dir = temp_path("dir");
        std::fs::create_dir(&dir).expect("create dir");
        let err = load_checks_from_yaml_file(&dir).unwrap_err();
        let _ = std::fs::remove_dir(&dir);
        assert!(
            format!("{err}").contains("not a regular file"),
            "got: {err}"
        );
    }

    #[cfg(unix)]
    #[test]
    fn rejects_symlink() {
        let target = temp_path("sym_target");
        let link = temp_path("sym_link");
        let _gt = TempFile(target.clone());
        let _gl = TempFile(link.clone());
        std::fs::write(&target, SAMPLE).expect("write target");
        if std::os::unix::fs::symlink(&target, &link).is_err() {
            return; // symlink creation not permitted — skip (mirrors the Go test)
        }
        let err = load_checks_from_yaml_file(&link).unwrap_err();
        assert!(format!("{err}").contains("symbolic link"), "got: {err}");
    }

    #[test]
    fn stat_failure_surfaces_errno_not_missing() {
        // A path component over NAME_MAX makes symlink_metadata (lstat) fail
        // with ENAMETOOLONG — a stat *failure*, not an absent file. The loader
        // embeds the raw io::Error ("cannot stat …: <errno>"), so a
        // resource/permission failure under load can never be mislabelled a
        // missing file (the cross-binding ec-vs-not-found invariant; cf. the
        // C++ validate_loader_path split and Go's os.ErrNotExist branch).
        // ENAMETOOLONG is deterministic and root-safe.
        let long = temp_path(&"a".repeat(5000));
        let err = load_checks_from_yaml_file(&long).unwrap_err();
        let msg = format!("{err}");
        assert!(msg.contains("cannot stat"), "got: {msg}");
        assert!(
            !msg.contains("No such file"),
            "stat failure mislabelled as a missing file: {msg}"
        );
    }
}
